go/plugins/weaviate/weaviate.go

```diff
@@ -0,0 +1,374 @@
+const provider = "weaviate"
+
+// The metadata key used to hold document text.
+const textKey = "text"
+
+// The metadata key to hold document metadata.
+const metadataKey = "metadata"
+
+// state holds the current plugin state.
+var state struct {
+	mu          sync.Mutex
+	initialized bool
+	client      *weaviate.Client
+}
+
+// getClient returns the client stored in the state.
+func getClient() *weaviate.Client {
+	state.mu.Lock()
+	defer state.mu.Unlock()
+	return state.client
+}
+
+// ClientConfig passes configuration options to the plugin.
+type ClientConfig struct {
+	// The hostname:port to use to contact the Weaviate database.
+	// If not set, the default is read from the WEAVIATE_URL environment variable.
+	Addr string
+	// The scheme to use to contact the Weaviate database.
+	// Typically http or https.
+	// If not set, the default is https.
+	Scheme string
+	// The API key to use with the Weaviate database.
+	// If not set, the default is read from the WEAVIATE_API environment variable.
+	APIKey string
+}
+
+// Init initializes the Weaviate plugin.
+// The cfg argument may be nil to use the defaults.
+// This returns the [*weaviate.Client] in case it is useful,
+// but many users will be able to use just [DefineIndexerAndRetriever].
+func Init(ctx context.Context, cfg *ClientConfig) (*weaviate.Client, error) {
+	state.mu.Lock()
+	defer state.mu.Unlock()
+	if state.initialized {
+		panic("weaviate.Init already called")
+	}
+
+	var host string
+	if cfg != nil {
+		host = cfg.Addr
+	}
+	if host == "" {
+		host = os.Getenv("WEAVIATE_URL")
+	}
+
+	var scheme string
+	if cfg != nil {
+		scheme = cfg.Scheme
+	}
+	if scheme == "" {
+		scheme = "https"
+	}
+
+	var apiKey string
+	if cfg != nil {
+		apiKey = cfg.APIKey
+	}
+	if apiKey == "" {
+		apiKey = os.Getenv("WEAVIATE_API")
+	}
+
+	config := weaviate.Config{
+		Host:       host,
+		Scheme:     scheme,
+		AuthConfig: auth.ApiKey{Value: apiKey},
+	}
+
+	client, err := weaviate.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("weaviate initialization failed: %v", err)
+	}
+
+	live, err := client.Misc().LiveChecker().Do(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("weaviate initialization failed: %v", err)
+	}
+	if !live {
+		return nil, errors.New("weaviate instance not alive")
+	}
+
+	state.client = client
+	state.initialized = true
+
+	return client, nil
+}
+
+// ClassConfig holds configuration options for an indexer/retriever pair.
+type ClassConfig struct {
+	// The weaviate class name. May not be the empty string.
+	Class string
+
+	// The Embedder and options to use to embed documents.
+	// Embedder may not be nil.
+	Embedder        ai.Embedder
+	EmbedderOptions any
+}
+
+// DefineIndexerAndRetriever defines an [ai.Indexer] and [ai.Retriever]
+// that use the same class.
+// The name uniquely identifies the Indexer and Retriever
+// in the registry.
+func DefineIndexerAndRetriever(ctx context.Context, cfg *ClassConfig) (ai.Indexer, ai.Retriever, error) {
```

Comment:
WDYT about the arg type being `ClassConfig` instead of `*ClassConfig`, since it's required?

Reply:
Done.